java/ql/lib/semmle/code/java/Type.qll

```diff
@@ -568,12 +568,30 @@ class RefType extends Type, Annotatable, Modifiable, @reftype {
     exists(Field f | f = m |
       f = this.getAField()
       or
-      not f.isPrivate() and not this.declaresField(f.getName()) and this.getASupertype().inherits(f)
+      not f.isPrivate() and
+      (f.isPackageProtected() implies f.getDeclaringType().getPackage() = this.getPackage()) and
+      not this.declaresField(f.getName()) and
+      this.getASupertype().inherits(f)
       or
       this.getSourceDeclaration().inherits(f)
     )
     or
     this.hasMethod(m, _)
+    or
+    // Constructors are not inherited, but this predicate also considers declared members
+    m.(Constructor).getDeclaringType() = this.getSourceDeclaration()
+    or
+    exists(MemberType memberType | memberType = m |
+      m.(MemberType).getDeclaringType() = this
+      or
+      not memberType.isPrivate() and
+      (memberType.isPackageProtected() implies memberType.getPackage() = this.getPackage()) and
+      // And member type is not hidden by another member type with the same name (regardless of visibility)
```

Comment:
There's also https://docs.oracle.com/javase/tutorial/java/IandI/hidevariables.html which would enable you to treat types and fields the same

Reply:
That seems to only be about fields though, isn't it? Do you mean fields could hide member types and member types could hide fields? I am not sure if that is possible, at least the JLS does not seem to mention it.

Or do you mean I should try to combine the common checks here for `Field` and `MemberType`?

Reply:
I meant fields hiding fields and types hiding types and forgot about the possibility of one hiding the other.